### PR TITLE
refactor: chunk KOS extend message

### LINKED ISSUE #97: chunk KOS extension message
The `Extend` message sent from the KOS receiver to the sender during extension can be very large. This clashes with DoS mitigation at the IO layer which limits the size of incoming data.

https://github.com/privacy-scaling-explorations/mpz/blob/b8f94795f7f33936a0cb97515408f0d4bb0789db/ot/mpz-ot/src/kos/receiver.rs#L108

Instead of sending it as a single message, the receiver should send it in chunks with a max size.

### PR DESCRIPTION
**Pull request by sinui0**
This PR refactors KOS impl to chunk the extension message sent from the Receiver to the Sender.

Closes #97 and https://github.com/tlsnotary/tlsn-utils/issues/10

### CODE REVIEW

**Review by th4s — COMMENTED** [review]
Wouldn't it be cleaner to handle this in a general way in our messaging layer? Then this change would apply to every sink and stream we are using, while at the same time not mixing the protocol code with IO stuff.

I am thinking of something like wrapper functions in our `IoSink` and `IoStream` traits, which break down the payload if needed.

**Comment by sinui0** [comment]
That thought certainly crossed my mind, but it didn't seem to be quite as straight forward.

1. The IO layer already breaks down the message into chunks. The limit we are running into is mostly regarding _deserialization_ and how much memory a received message is allowed to utilize. This limit needs to be specific to the application context and message type. We could add additional bounds to the IO traits to specify this, but that would make them harder to implement and use. This would be less of a problem with specialization, but that's not in the foreseeable future.
2. A less flexible middleground is to fork `tokio_util::codec::length_delimited::LengthDelimitedCodec` and make it aware of variant-level size limits, so we could specify a limit specifically for `Extend` at the codec level rather than baking this into the IO traits. I don't want to block increasing the TLSNotary protocol limit on this work.
3. I think at the end of the day the chunking will get baked into the callsite control flow anyhow. The optimal state of this code would be to *generate* the matrix in chunks, streaming them, and processing the chunks concurrently. That would require "chunk" message types anyhow. Given our deprecation plans, this is a simpler short-term remedy even if its not actually doing stream processing.

**Comment by th4s** [comment]
Makes sense :+1:, I haven't put that much thought into it when I wrote that.

**Review by th4s — APPROVED** [review]
:+1:

File: ot/mpz-ot-core/src/kos/msgs.rs [hunk 1]
```diff
@@ -33,15 +34,55 @@ impl<BaseMsg> From<MessageError<BaseMsg>> for std::io::Error {
...
+
+/// Extension message sent by the receiver.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct Extend {
     /// The receiver's extension vectors.
     pub us: Vec<u8>,
 }
 
+impl Extend {
+    /// Splits the message into chunks.
```
Comment (themighty1) [inline ot/mpz-ot-core/src/kos/msgs.rs]:
```suggestion
    /// Returns an iterator over the chunks of the message.
```

File: ot/mpz-ot-core/src/kos/mod.rs [hunk 2]
```diff
@@ -28,6 +28,11 @@ pub(crate) type RngSeed = <Rng as SeedableRng>::Seed;
 /// AES-128 CTR used for encryption.
 pub(crate) type Aes128Ctr = ctr::Ctr64LE<aes::Aes128>;
 
+/// Returns the size in bytes of the extension message for a given number of OTs.
```
Comment (themighty1) [inline ot/mpz-ot-core/src/kos/mod.rs]:
```suggestion
/// Returns the size in bytes of the extension matrix for a given number of OTs.
```

File: ot/mpz-ot/src/kos/sender.rs [hunk 3]
```diff
@@ -138,13 +140,36 @@ where
...
+
+        // Receive extension matrix from the receiver.
+        while extend.us.len() < expected_us {
+            let Extend { us: chunk } = stream
+                .expect_next()
+                .await?
+                .try_into_extend()
+                .map_err(SenderError::from)?;
+
+            extend.us.extend(chunk);
```
Comment (themighty1) [inline ot/mpz-ot/src/kos/sender.rs:172]: Is this correct?
The receiver was splitting off from the end of the matrix, so the first chunk which arrives must go to the end of the reconstructed matrix, not to the beginning as it is here.

**Review by themighty1 — CHANGES_REQUESTED** [review]

File: ot/mpz-ot/src/kos/sender.rs (hunk 3 above)
Comment (sinui0) [reply ot/mpz-ot/src/kos/sender.rs:172]: Woh, what... how do the unit tests pass? I'll look into it

Comment (sinui0) [reply ot/mpz-ot/src/kos/sender.rs:172]: Good catch, the unit tests don't setup enough OTs to hit the chunk threshold.

Comment (sinui0) [reply ot/mpz-ot/src/kos/sender.rs:172]: Fixed.

File: ot/mpz-ot/src/kos/receiver.rs [hunk 4]
```diff
@@ -90,9 +93,11 @@ where
         let mut ext_receiver =
             std::mem::replace(&mut self.state, State::Error).try_into_extension()?;
 
+        let count = pad_ot_count(count);
+
         // Extend the OTs, adding padding for the consistency check.
         let (mut ext_receiver, extend) = Backend::spawn(move || {
-            let extend = ext_receiver.extend(count + CSP + SSP);
+            let extend = ext_receiver.extend(count);
```
Comment (themighty1) [inline ot/mpz-ot/src/kos/receiver.rs:100]: padding should not happen inside .extend() again

File: ot/mpz-ot/src/kos/receiver.rs [hunk 5]
```diff
@@ -90,9 +93,11 @@ where
         let mut ext_receiver =
             std::mem::replace(&mut self.state, State::Error).try_into_extension()?;
 
+        let count = pad_ot_count(count);
+
         // Extend the OTs, adding padding for the consistency check.
```
Comment (themighty1) [inline ot/mpz-ot/src/kos/receiver.rs]: this comment is not accurate anymore

**Review by themighty1 — CHANGES_REQUESTED** [review]

File: ot/mpz-ot/src/kos/receiver.rs (hunk 4 above)
Comment (sinui0) [reply ot/mpz-ot/src/kos/receiver.rs:100]: I've removed the rounding that occurs in the core impl and replaced it with an input restriction to be a multiple of 64. This subtlety is not exposed to the user of the async wrapper.

**Review by themighty1 — APPROVED** [review]
gw

**Pull request merged**